scheduler/scd_utils.py

```diff
@@ -0,0 +1,239 @@
+        self.scd_filename = scd_filename
+        self.scd_dt_fmt = "%x %H:%M"
+        self.line_fmt = "{datetime} {duration} {prio} {experiment}"
+
+    def read_scd(self):
+        """Read and parse the SCD file.
+
+        Returns:
+            TYPE: Dict
+
+        Raises:
+            ValueError: on lines with obvious errors in them.
+            Will also throw file errors if there are problems opening SCD file.
+        """
+        with open(self.scd_filename, "r") as f:
+            raw_scd = f.readlines()
+
+        raw_scd = [line.split() for line in raw_scd]
+
+        scd_lines = []
+        for num, line in enumerate(raw_scd):
+            if len(line) != 5:
+                raise ValueError("Line {} has too many arguments".format(num))
+
+            try:
+                time = dt.datetime.strptime(line[0] + " " + line[1], self.scd_dt_fmt)
+            except:
+                raise ValueError("Unable to create datetime from line {}".format(num))
+
+            duration = line[2]
+
+            prio = int(line[3])
+            if not (0 <= prio <= 20):
+                raise ValueError("Priority in line {} is out of bounds".format(num))
+
+            experiment = line[4]
+
+            epoch = dt.datetime.utcfromtimestamp(0)
+            epoch_milliseconds = int((time - epoch).total_seconds() * 1000)
+
+            scd_lines.append({"timestamp" : epoch_milliseconds,
+                                "time" : time,
+                                "duration" : str(duration),
+                                "prio" : str(prio),
+                                "experiment" : experiment})
+
+        return scd_lines
+
+    def fmt_line(self, line_dict):
+        """Formats a dictionary with line info into a text line for file.
+
+        Args:
+            line_dict (TYPE): A dict that holds all the line info.
+
+        Returns:
+            TYPE: Formatted string.
+        """
+        line_str = self.line_fmt.format(datetime=line_dict["time"].strftime(self.scd_dt_fmt),
+                                        prio=line_dict["prio"],
+                                        experiment=line_dict["experiment"],
+                                        duration=line_dict["duration"])
+        return line_str
+
+    def write_scd(self, scd_lines):
+        """Creates SCD text lines and writes to file. Creates a backup of the old file before
+        writing.
+
+        Args:
+            scd_lines (TYPE): A list dicts that contain the line info.
+        """
+        text_lines = [self.fmt_line(x) for x in scd_lines]
+
+        shutil.copy(self.scd_filename, self.scd_filename+".bak")
+
+        with open(self.scd_filename, 'w') as f:
+            for line in text_lines:
+                f.write("{}\n".format(line))
+
+
+    def add_line(self, mmddyy, hhmm, prio, experiment, duration='-'):
+        """Adds a new line to the SCD.
+
+        Args:
+            mmddyy (TYPE): month/day/year string.
```

Comment:
same comment as above - I think it would be nice to standardize to YYYYMMDD to keep things the same from data files to schedule lines. Thoughts?

Reply:
changed it